zipline/pipeline/loaders/events.py

```diff
@@ -0,0 +1,154 @@
+        occurred along with other associated values.
+
+        If the DataFrames contain a "timestamp" column, that column is
+        interpreted as the date on which we learned about the event.
+
+        If the DataFrames do not contain a "timestamp" column, we assume we
+        knew about the event on all prior dates.  This mode is only supported
+        if ``infer_timestamp`` is explicitly passed as a truthy value.
+    infer_timestamps : bool, optional
+        Whether to allow omitting the "timestamp" column.
+    dataset : DataSet
+        The DataSet object for which this loader loads data.
+    expected_cols : frozenset
+        Set of expected columns for the dataset, without timestamp.
+    """
+
+    def __init__(self,
+                 all_dates,
+                 events_by_sid,
+                 infer_timestamps=False,
+                 dataset=None,
+                 expected_cols=frozenset()):
+        self.all_dates = all_dates
+        # Do not modify the original in place, since it may be used for other
+        #  purposes.
+        self.events_by_sid = (
+            events_by_sid.copy()
+        )
+        dates = self.all_dates.values
+
+        for k, v in iteritems(events_by_sid):
+            # First, must convert to DataFrame.
+            if isinstance(v, pd.Series):
+                # If Series was passed, DateTime index is assumed.
+                self.events_by_sid[k] = pd.DataFrame(v)
+            elif isinstance(v, pd.DatetimeIndex):
+                if not infer_timestamps:
+                    raise ValueError(
+                        "Got DatetimeIndex for sid %d.\n"
+                        "Pass `infer_timestamps=True` to use the first date in"
+                        " `all_dates` as implicit timestamp." % k
+                    )
+                self.events_by_sid[k] = pd.DataFrame(v)
+                v.index = [dates[0]] * len(v)
+            # Already a DataFrame
+            elif isinstance(v, pd.DataFrame):
+                if TS_FIELD_NAME not in v.columns:
+                    if not infer_timestamps:
+                        raise ValueError(
+                            "Got DataFrame without a '%s' column for sid %d.\n"
+                            "Pass `infer_timestamps=True` to use the first "
+                            "date in `all_dates` as implicit timestamp." %
+                            (TS_FIELD_NAME, k)
+                        )
+                    self.events_by_sid[k] = v = v.copy()
+                    v.index = [dates[0]] * len(v)
+                else:
+                    self.events_by_sid[k] = v.set_index(TS_FIELD_NAME)
+            else:
+                raise ValueError("Data for sid %s must be in DataFrame, "
+                                 "Series, or DatetimeIndex." % k)
+            # Once data is in a DF, make sure columns are correct.
+            cols_except_ts = (set(v.columns.values) -
```

Comment:
we shouldn't need to thake the values of the columns here, this will coerce to an ndarray of strings. Also, why are we stripping out a sid column?

Do you think we should union the `TS_FIELD_NAME` with the expected fields earlier so that at runtime a user can inspect what the expected fields are and see that they need to iinclude `TS_FIELD_NAME`